aten/src/ATen/native/Sorting.cpp

```diff
@@ -183,6 +183,137 @@ std::tuple<Tensor, Tensor> kthvalue(
+}
+
+Tensor& quantile_out(
+    Tensor& out,
+    const Tensor& self,
+    const Tensor& q,
+    c10::optional<int64_t> _dim,
+    bool keepdim) {
+  auto out_shape = quantile_out_shape(self, q, _dim, keepdim);
+  TORCH_CHECK(out.sizes().vec() == out_shape, 
+      "expected out shape to be ", out_shape, " but got ", out.sizes().vec());
+
+  TORCH_CHECK(self.numel() > 0, "Input tensor must be non-empty");
+  TORCH_CHECK(q.dim() <= 1, "q must be a scalar or 1D tensor");
+  
+  TORCH_CHECK(self.scalar_type() == kFloat || self.scalar_type() == kDouble, 
+      "Input tensor must be either float or double dtype");
+  TORCH_CHECK(self.scalar_type() == q.scalar_type(), 
+      "q must be same dtype as the input tensor");
+  TORCH_CHECK(self.scalar_type() == out.scalar_type(), 
+      "out tensor must be same dtype as the input tensor");
+  
+  TORCH_CHECK(self.device() == q.device(), 
+      "q must be on the same device as the input tensor");
+  TORCH_CHECK(self.device() == out.device(), 
+      "out tensor must be on the same device as the input tensor");
+
+  if (q.numel() == 0) {
+    return out;
+  }
+
+  if (self.device() == kCPU) {
+    TORCH_CHECK(q.ge(0).logical_and_(q.le(1)).all().item<bool>(),
+        "q values must be in the range [0, 1]");
+  }
+
+  // If user didn't specify a dimension then we flatten the input tensor
+  auto in = _dim ? self : self.flatten();
+  auto dim = _dim ? at::maybe_wrap_dim(*_dim, self.dim(), true) : 0;
+
+  // We sort the input tensor to efficiently query multiple kth values.
+  // In the future, it might be worthwhile to implement multiple quickselect.
+  auto sorted = std::get<0>(in.sort(dim));
+
+  // Convert quantile into indices in the given dimension.
+  // Check for overflow when casting to a floating point type.
+  TORCH_CHECK(sorted.size(dim) <= std::pow(2, 24), "Input tensor is too large");
+  auto indices = q * (sorted.size(dim) - 1);
+  auto indices_below = indices.floor().toType(kLong);
+  auto indices_above = indices.ceil().toType(kLong);
+
+  // Extract values from tensor
+  auto values_below = sorted.index_select(dim, indices_below);
+  auto values_above = sorted.index_select(dim, indices_above);
+
+  // Interpolate values to get quantiles
+  std::vector<int64_t> broadcast_shape(sorted.dim(), 1);
+  broadcast_shape[dim] = -1;
+  auto weights = (indices - indices_below).reshape(broadcast_shape);
+  auto quantiles = values_below.lerp_(values_above, weights);
+
+  // when the q tensor is not a scalar, numpy will prepend a new dimension
+  // to contain the quantiles. This code ensures we follow the same order
+  if (q.dim() > 0) {
+    quantiles.unsqueeze_(0);
+    ++dim;
+    std::vector<int64_t> numpy_dim_order;
+    for (int64_t i = 0; i < quantiles.dim(); ++i) {
+      numpy_dim_order.push_back(i);
+    }
+    std::iter_swap(numpy_dim_order.begin(), numpy_dim_order.begin() + dim);
+    quantiles = quantiles.permute(numpy_dim_order);
+  }
+
+  out.copy_(quantiles.reshape(out_shape));
```

Comment:
This copy is unfortunate. Can it be eliminated? One way to eliminate it most of the time would be to have this function return the reshaped quantiles tensor and have the out variant perform the shape check and copy it into the out_ param, if given. Another option would be to not support an out variant at all. Out is little-used and getting the result shape of this operation correct is tricky, after all.

Reply:
I'll make it so only the out variant makes the copy but I agree that this is not optimal. I also don't like how I have to permute the dimensions to follow NumPy shape. I think this would really benefit from a custom cpu/cuda implementation. Maybe a follow-up PR?